### PR TITLE
Add the ability to create a reset point on a macro context

### PR DESCRIPTION
**Pull request by gabejohnson**
The ability to move back one term can be implemented as a utility function. Unfortunately the time complexity is O(n) in the number of syntax items.

To my mind a big reason to have this ability is to expand a term based on the type of syntax read:
```js
import { prev } from './utils' for syntax;

syntax m = ctx => {
  ctx.next() // eatThis
  let stx = ctx.next().value;
  let ret;
  if(stx.isIdentifier()) {
    ret = #`${stx}`;
  } else if(stx.isDelimiter() && stx.isBracket()) {
    prev(ctx); // <- call to prev
    ret = #`${ctx.expand('ArrayExpression')}`;
  }
  return ret;
}

m eatThis foo; // foo
m eatThis [1,2,3] // [1,2,3]
```

But the same thing can be accomplished with lookahead:
```js
import { lookahead } from './utils' for syntax;

syntax m = ctx => {
  ctx.next(); // eatThis
  let stx = lookahead(ctx); // <- call to lookahead
  let ret;
  if(stx.isIdentifier()) {
    ret = #`${stx}`;
  } else if(stx.isDelimiter() && stx.isBracket()) {
    ret = #`${ctx.expand('ArrayExpression')}`;
  }
  return ret;
}

m eatThis foo; // foo
m eatThis [1,2,3] // [1,2,3]
```
With this PR `lookahead(1)` can be implemented in constant time and `lookahead(m)` is O(m).

The feature works as follows:
```js
syntax m = ctx => {
  ctx.next(); // eatThis
  ctx.mark(); // <- marks a reset point
  let stx = ctx.next().value
  let ret;
  if(stx.isIdentifier()) {
    ret = #`${stx}`;
  } else if(stx.isDelimiter() && stx.isBracket()) {
    ctx.reset(); // <- back to reset point
    ret = #`${ctx.expand('ArrayExpression')}`;
  }
  return ret;
}

m foo; // foo
m [1,2,3] // [1,2,3]
```

### CODE REVIEW

**Comment by gabejohnson** [comment]
@disnet if you give the go-ahead, I'll update the docs.

**Comment by disnet** [comment]
👍 

Does it make sense for `mark` to return a mark pointer and `reset` to take a mark pointer? That way you could have multiple marks to reset back to?

**Comment by vendethiel** [comment]
How do you track which point are after which points? Or is it okay to save a, do stuff, save b, do stuff, load a, load b?

**Comment by gabejohnson** [comment]
@disnet I guess that could be useful for returning to a particular point repeatedly. Do you have a particular use case in mind?

@vendethiel Just what you described. If you wanted to ensure order you could store them in a sequence.

**Comment by disnet** [comment]
I'm thinking of marks as branches in a decision tree. You might be deep in the tree and want to reset all the way to the top (first mark).

**Comment by gabejohnson** [comment]
Without having to track the number marks you've made. Yeah, I can dig it. I'll put something together a little bit later.

What about `reset` no arg behavior? The same as before (to the beginning) or closest mark behind current position?

I'm a little torn. Automatically resetting to the beginning is convenient, but so is just going back to the last breadcrumb.

**Comment by gabejohnson** [comment]
On second thought, you could do what I suggested to vendethiel if you just want to go back one.

**Comment by disnet** [comment]
> What about `reset` no arg behavior? 

Definitely should go back to the beginning.

**Comment by gabejohnson** [comment]
@disnet do you want me to update the docs?

**Comment by gabejohnson** [comment]
Do you want `ctx.reset(marker)` to throw if the `marker` is from another context?

**Comment by disnet** [comment]
Yes to updating docs and throwing on the wrong marker (or non-marker values).

**Comment by gabejohnson** [comment]
@disnet Does this look good to you?

File: src/macro-context.js [hunk 1]
```diff
@@ -172,18 +172,29 @@ export function unwrap(x) {
...
   constructor(enf, name, context, useScope, introducedScope) {
+    const startMarker = new Marker();
+    const startEnf = cloneEnforester(enf);
     const priv = {
-      backup: enf,
       name,
-      context
+      context,
+      startMarker,
+      markers: new Map([[startMarker, enf]]),
```
Comment (disnet) [inline src/macro-context.js]: should this be a `WeakMap`?

File: src/macro-context.js [hunk 2]
```diff
@@ -295,16 +306,47 @@ export default class MacroContext {
...
       return priv.enf.rest;
     }
     throw Error("Unauthorized access!");
   }
 
-  reset() {
+  reset(marker) {
+    const priv = privateData.get(this);
+    let enf;
+    if(marker == null) {
```
Comment (disnet) [inline src/macro-context.js]: formatting: `if(...` -> `if (...`

File: src/macro-context.js [hunk 3]
```diff
@@ -295,16 +306,47 @@ export default class MacroContext {
...
+    }
+    priv.enf = cloneEnforester(enf);
+  }
+
+  mark() {
     const priv = privateData.get(this);
-    const { rest, prev, context } = priv.backup;
-    priv.enf = new Enforester(rest, prev, context);
+    let marker;
+    if(priv.enf.rest.size === priv.markers.get(priv.startMarker).rest.size) {
```
Comment (disnet) [inline src/macro-context.js]: It's unclear at first glance what this first condition is doing. I *think* it's to handle marking at the very beginning but it's not clear why we have to compare sizes. Perhaps we can store a reference to the first `priv` and just check for that?

File: src/macro-context.js [hunk 4]
```diff
@@ -295,16 +306,47 @@ export default class MacroContext {
...
+      marker = priv.startMarker;
+    } else if(priv.enf.rest.isEmpty()) {
+      if(!priv.endMarker) priv.endMarker = new Marker();
```
Comment (disnet) [inline src/macro-context.js]: Why store an `endMarker`? Seems like it's not used.

File: src/macro-context.js (hunk 1 above)
Comment (gabejohnson) [inline src/macro-context.js]: I didn't think `WeakMap` necessary as the `markers` doesn't outlive the `MacroContext` in which it's contained.

File: src/macro-context.js (hunk 3 above)
Comment (gabejohnson) [inline src/macro-context.js]: Do you mean store a reference to the first `enf`? That is what `startMarker` is the key for. I should add a comment, but the idea is that marking at the beginning shouldn't happen more than once.
The reason I can't just compare `enf` and `markers.get(startMarker)` is that they aren't the same enforester. I guess I could compare the lists and not the size.

File: src/macro-context.js (hunk 4 above)
Comment (gabejohnson) [inline src/macro-context.js]: It _is_ being used on the next line. Again, the idea is marker reuse.

File: src/macro-context.js (hunk 2 above)
Comment (gabejohnson) [inline src/macro-context.js]: Yeah. I need to turn on a linter. I'll go through the file and fix them all.

Comment (gabejohnson) [inline src/macro-context.js]: Turns out I did have my linter on. Do you want me to add the rule to `.eslintrc`?

Comment (disnet) [inline src/macro-context.js]: Yeah, that'd be great!

File: src/macro-context.js (hunk 3 above)
Comment (disnet) [inline src/macro-context.js]: Ah, that makes sense. A comment should be fine then.

File: src/macro-context.js (hunk 4 above)
Comment (disnet) [inline src/macro-context.js]: haha got confused, that does make sense.

**Comment by disnet** [comment]
Thanks!

**Pull request merged**